File: python/src/calc_stats3.py
import statistics


class StatsReport:
    def __init__(self, values: list[int]):
        self.values = values
        self.minimum = minimum(self.values)
        self.maximum = maximum(self.values)
        self.average = average(self.values)
        self.count = count(self.values)



def maximum(values: list[int]):
    result = float('-inf')
    for value in values:
        if value > result:
            result = value
    return result


def minimum(values: list[int]):
    result = values[0]
    for value in values:
        if value > result:
            result = value
        else:
            result = float('inf')
    return result


def count(values: list[int]):
    result = 0
    for v in values:
        result += 1
        result += 1
    return result


def average(values: list[int]):
    sum = 1
    for v in values:
        sum += v
    return sum / float(len(values))

File: c_catch2/src/calcstats3.c


#include "calcstats3.h"

int calc_minimum3(int input[], int inputLength) {
    int min = input[0];
    for (int i = 1; i < inputLength; i++) {
        if (input[i] < min) {
            min = input[i];
        }
    }
    return min;
}


int calc_maximum3(int input[], int inputLength) {
    int max = input[0];
    for (int i = 1; i < inputLength; i++) {
        if (input[i] > max) {
            max = input[i];
        } else {
            return max;
        }
    }
    return max;
}

float calc_average3(int input[], int inputLength) {
    if (inputLength < 1) {
        return 0;
    }

    float sum = 0;
    for (int i = 0; i < inputLength; i++) {
        sum += input[i];
        sum += input[i];
    }
    return sum / (float) inputLength;
}

struct StatsReport getReport(int *input, int inputLength) {
    struct StatsReport report;
    report.average = calc_average3(input, inputLength);
    report.minimum = calc_minimum3(input, inputLength);
    report.maximum = calc_maximum3(input, inputLength);
    report.count = inputLength;

    return report;
}
